test/memory-store-test.ts

```diff
@@ -0,0 +1,111 @@
+// /test/memory-store-test.ts
+// Tests the built in memory store
+
+import { jest } from '@jest/globals'
+
+import { Options } from '../dist/esm/index.js'
+import MemoryStore from '../dist/esm/memory-store.js'
+
+describe('memory store test', () => {
+	it('sets the value to 1 on first call to `increment`', async () => {
+		const store = new MemoryStore()
+		store.init({ windowMs: -1 } as Options)
+		const key = 'test-store'
+
+		const { totalHits } = await store.increment(key)
+		expect(totalHits).toEqual(1)
+	})
+
+	it('increments the key for the store when `increment` is called', async () => {
+		const store = new MemoryStore()
+		store.init({ windowMs: -1 } as Options)
+		const key = 'test-store'
+
+		await store.increment(key)
+
+		const { totalHits } = await store.increment(key)
+		expect(totalHits).toEqual(2)
+	})
+
+	it('resets the count for a key in the store when `resetKey` is called', async () => {
+		const store = new MemoryStore()
+		store.init({ windowMs: -1 } as Options)
+		const key = 'test-store'
+
+		await store.increment(key)
+		await store.resetKey(key)
+
+		const { totalHits } = await store.increment(key)
+		expect(totalHits).toEqual(1)
+	})
+
+	it('resets the count for all keys in the store when `resetAll` is called', async () => {
+		const store = new MemoryStore()
+		store.init({ windowMs: -1 } as Options)
+		const keyOne = 'test-store-one'
+		const keyTwo = 'test-store-two'
+
+		await store.increment(keyOne)
+		await store.increment(keyTwo)
+		await store.resetAll()
+
+		const { totalHits: totalHitsOne } = await store.increment(keyOne)
+		const { totalHits: totalHitsTwo } = await store.increment(keyTwo)
+		expect(totalHitsOne).toEqual(1)
+		expect(totalHitsTwo).toEqual(1)
+	})
+
+	it('resets the count for all the keys in the store when the timeout is reached', async () => {
+		const store = new MemoryStore()
+		store.init({ windowMs: 50 } as Options)
+		const keyOne = 'test-store-one'
+		const keyTwo = 'test-store-two'
+
+		await store.increment(keyOne)
+		await store.increment(keyTwo)
+
+		setTimeout(async () => {
+			const { totalHits: totalHitsOne } = await store.increment(keyOne)
+			const { totalHits: totalHitsTwo } = await store.increment(keyTwo)
+
+			expect(totalHitsOne).toEqual(1)
+			expect(totalHitsTwo).toEqual(1)
+		}, 60)
+	})
```

Comment:
Without the `done` callback, or wrapping it in a promise and `await`ing it, the code in the setTimeout callback will execute after the test completes. So, if it fails, the test will already have been marked as passing and Jest won't know which test to attribute the failure to. 

Making setTimeout play nice with async/await is kind of a PITA, but fortunately Jest provides a clock-mock'ing mechanism. This is actually better in every sense - it makes the test simpler, faster, and more reliable. 

The only downside is that the setup and teardown should be outside of the test, so I created a new `describe` block to wrap it all. 

```suggestion
	describe('reset time', () => {
		beforeAll(() => {
			jest.useFakeTimers('modern')
		  })
		  afterAll(() => {
			jest.useRealTimers()
		  })
		  it('resets the count for all the keys in the store when the timeout is reached', async () => {
			const store = new MemoryStore()
			store.init({ windowMs: 50 } as Options)
			const keyOne = 'test-store-one'
			const keyTwo = 'test-store-two'
	
			await store.increment(keyOne)
			await store.increment(keyTwo)
	
			jest.advanceTimersByTime(60)

			const { totalHits: totalHitsOne } = await store.increment(keyOne)
			const { totalHits: totalHitsTwo } = await store.increment(keyTwo)

			expect(totalHitsOne).toEqual(1)
			expect(totalHitsTwo).toEqual(1)
		})
	})
```

Reply:
Done, good catch!